sycl/doc/extensions/proposed/sycl_ext_intel_low_power_event.asciidoc

```diff
@@ -0,0 +1,153 @@
+
+
+== Overview
+
+On some backends, calling `wait()` on an `event` will synchronize using a
+busy-waiting implementation. Though this comes at a low latency for the
+synchronization of the event, it has the downside of consuming high amounts of
+CPU time for no meaningful work.  This extension introduces a new property for
+SYCL commands that will produce a "low-power" event. These new low-power events
+will, if possible, yield the thread that the `wait()` member function is called
+on and only wake up occasionally to check if the event has finished. This
+reduces the time the CPU spends checking finish condition of the wait, at the
+cost of latency.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_LOW_POWER_EVENT` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Low-power event property
+
+This extension adds a new property which can be used with the
+`submit_with_event` free function from
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[sycl_ext_oneapi_enqueue_functions].
+Passing this property to this function will act as a hint to the `event` created
+from the corresponding commands to do low-power synchronization. If the backend
+is able to handle low-power events, calling `event::wait()` or
+`event::wait_and_throw()` will cause the thread to yield and only do occasional
+wake-ups to check the event progress.
+
+[_Note:_ The property currently only has an effect on `barrier` and
+`partial_barrier` commands enqueued on queues that return
+`backend::ext_oneapi_level_zero` from `queue::get_backend()`.
+_{endnote}_]
+
+```
+namespace sycl::ext::intel::experimental {
+
+struct low_power_event_key {
+  using value_t = 
+      oneapi::experimental::property_value<low_power_event_key>;
+};
+
+inline constexpr low_power_event_key::value_t low_power_event;
+
+} // namespace sycl::ext::intel::experimental
+```
+
+=== New property usage example
+
+As an example of how to use the new `low_power_event` property, see the
+following code:
+
+```
+#include <sycl/sycl.hpp>
+
+namespace oneapiex = sycl::ext::oneapi::experimental;
+namespace intelex = sycl::ext::intel::experimental;
+
+int main() {
+  sycl::queue Q;
+
+  // Submit some work to the queue.
+  oneapiex::submit(Q, [&](sycl::handler &CGH) {...});
+
+  // Submit a command with the low-power event property.
+  oneapiex::properties Props{intelex::low_power_event};
```

Comment:
Presently, our `wait` calls end up in the UR which calls `zeEventHostSynchronize` on L0.  ( https://spec.oneapi.io/level-zero/0.95/core/api.html#zeeventhostsynchronize ).  That call takes a nanoseconds arg, which is how long it block before reporting back not completed.  Presently the UR passes `UINT64_MAX` which makes it block fully until the operation completes.  Is there a different API call planned to be used?

My worry with this design is that we are handwaving and saying "do low power", but what is that, exactly? Personally, I think the user should provide the sleep timeout. That way they are in control of how much accuracy they want to sacrifice for how much less CPU usage.

Reply:
nm, apparently the L0 call will behave differently for low power zeEvents.